src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubMessage.java

```diff
@@ -0,0 +1,179 @@
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Wrapper class for Pub/Sub message contents.
+ *
+ * We use this class to mirror the contents of Google's com.google.pubsub.v1.ReceivedMessage without having to add
+ * an additional exported dependency for our workers.
+ */
+public class PubSubMessage implements Externalizable {
+
+  private static final long serialVersionUID = 1L;
+
+  private String messageId;
+  private String orderingKey;
+  private String ackId;
+  private byte[] data;
+  private Map<String, String> attributes;
+  private Instant publishTime;
+
+  @VisibleForTesting
+  public PubSubMessage() {
+  }
+
+  @VisibleForTesting
+  public PubSubMessage(String messageId, String orderingKey, String ackId, byte[] data,
+                       Map<String, String> attributes, Instant publishTime) {
+    this.messageId = messageId;
+    this.orderingKey = orderingKey;
+    this.ackId = ackId;
+    this.data = data;
+    this.attributes = attributes;
+    this.publishTime = publishTime;
+  }
+
+  public PubSubMessage(ReceivedMessage message) {
+    if (message == null) {
+      return;
+    }
+
+    if (message.getMessage() != null) {
+      this.messageId = message.getMessage().getMessageId();
+      this.orderingKey = message.getMessage().getOrderingKey();
+
+      if (message.getMessage().getData() != null) {
+        this.data = message.getMessage().getData().toByteArray();
+      }
+
+      if (message.getMessage().getAttributesMap() != null) {
+        this.attributes = new HashMap<>(message.getMessage().getAttributesMap());
+      }
+
+      if (message.getMessage().getPublishTime() != null) {
+        this.publishTime = Instant.ofEpochSecond(message.getMessage().getPublishTime().getSeconds())
+          .plusNanos(message.getMessage().getPublishTime().getNanos());
+      }
+    }
+
+    this.ackId = message.getAckId();
+  }
+
+  public String getMessageId() {
+    return messageId;
+  }
+
+  public String getOrderingKey() {
+    return orderingKey;
+  }
+
+  public String getAckId() {
+    return ackId;
+  }
+
+  public byte[] getData() {
+    return data;
+  }
+
+  public Map<String, String> getAttributes() {
+    return attributes;
+  }
+
+  public Instant getPublishTime() {
+    return publishTime;
+  }
+
+  @Override
+  public void writeExternal(ObjectOutput out) throws IOException {
+    out.writeInt(messageId != null ? messageId.length() : -1);
+    if (messageId != null) {
+      out.writeUTF(messageId);
+    }
+
+    out.writeInt(orderingKey != null ? orderingKey.length() : -1);
+    if (orderingKey != null) {
+      out.writeUTF(orderingKey);
+    }
+
+    out.writeInt(ackId != null ? ackId.length() : -1);
+    if (ackId != null) {
+      out.writeUTF(ackId);
+    }
+
+    out.writeObject(publishTime);
+
+    out.writeInt(attributes != null ? attributes.size() : -1);
+    if (attributes != null) {
+      for (Map.Entry<String, String> entry : attributes.entrySet()) {
+        out.writeObject(entry.getKey());
```

Comment:
should this be writeUTF?

Reply:
Since the backing structure is a HashMap, null keys and null values are a possibility. `writeUTF` does not support null strings.

Reply:
Since you have many cases that need to write out a nullable String, introduce a private method for that purpose (write a boolean and then the non-null string as I commented above).

Reply:
I will, thanks for the advice @chtyim